### ISSUE TITLE
Vault sys/wrapping/unwrap api returns null data instead of original key-values

### ISSUE THREAD

**Issue by rajanadar** [hashicorp/vault#1983]
Vault Version 0.6.2

**Repro Steps:**

1. use the 'sys/wrapping/wrap' Vault Api to write a a list of key-values to Vault with wrap ttl as 1h.
2. the response has the expected wrap_info with the temp wrap token, ttl and creation_time. all good.
3. now use the temp wrap token, to invoke the 'sys/wrapping/unwrap' api. (POST call with 'token' body parameter) 

**Observed Result:**

1. the Data field in the returned response is null.

**Expected Result:**

1. the original list of key value pairs.

Instead of 'sys/wrapping/unwrap', if we use the deprecated 'cubbyhole/response' read, then i can see the original key value pairs as expected. i.e. data, along with a warning to use the /unwrap api. (as expected)

am i missing something?

**Comment by jefferai** [comment]
You're missing a key part in your observed results: it's not that the data field is null, it's that you're getting back an error saying that the token is invalid.

If you're using the wrapping token as the client token, don't pass in a `token` parameter to the unwrapping endpoint. In order to prevent multiple uses the token is expired before the results of the request are generated, so if you use the wrapping token to make the call but also pass it as the `token` parameter it will attempt to look up the wrapping token and see that it has already been revoked.

**Comment by rajanadar** [comment]
I should clarify. I didn't use the wrapping token as the client token for the unwrap api call. I used a separate root token. And I passed the wrapping token in the request body. 

And in the unwrap api response,  I don't see any error or warning. I actually see the wrap_info populated with the ttl,   creation time etc. Just that the data field is null. 

So in order to verify if data was really cubby holed for the wrap token,  I made a fresh wrap api call to get a new wrap token (which will have 1 count left)  and then made a call to the deprecated cubbyhole/ with the client token as the wrap token. (no parameters)  and I saw my original data in the data field. It means the way I am invoking the sys/wrapping/wrap api is really cubby holing the data. Just that the unwrap api did not return it. If I make another unwrap call immediately again, I see the token invalid or missing error. Which is totally expected. 

In summary,  here are the overall steps. Use a root client token to make a wrap api call with some data and the wrap header, and get a wrap token back. 
Make another root client token call to the unwrap api with the the wrap token as the body parameter. I get the response with wrap_info populated nicely,  but the data is null. 

Please let me know if I am missing anything.

**Comment by jefferai** [comment]
I can't repro:
```
$ vault write sys/wrapping/wrap foo=bar
Key                             Value
---                             -----
wrapping_token:                 4d7c6093-1162-b62f-36cf-d9729f25b6b6
wrapping_token_ttl:             5m0s
wrapping_token_creation_time:   2016-10-09 17:31:22.480296657 -0400 EDT

$ vault unwrap -format=json 4d7c6093-1162-b62f-36cf-d9729f25b6b6
{
        "request_id": "cf96de23-8a65-f3e9-6bb5-db986756635b",
        "lease_id": "",
        "lease_duration": 0,
        "renewable": false,
        "data": {
                "foo": "bar"
        },
        "warnings": null
}
```

**Comment by jefferai** [comment]
I can't repo using `sys/wrapping/unwrap` directly either:
```
$ vault write sys/wrapping/wrap foo=bar
Key                             Value
---                             -----
wrapping_token:                 5d4a9854-16a0-06ab-a8c9-248483522f05
wrapping_token_ttl:             5m0s
wrapping_token_creation_time:   2016-10-09 17:32:47.780244073 -0400 EDT

$ vault write -format=json sys/wrapping/unwrap token=5d4a9854-16a0-06ab-a8c9-248483522f05
{
        "request_id": "3ef7f3e2-01e7-6acb-a7b0-fb8a0d90f5b4",
        "lease_id": "",
        "lease_duration": 0,
        "renewable": false,
        "data": {
                "foo": "bar"
        },
        "warnings": null
}
```

**Comment by rajanadar** [comment]
thanks @jefferai for your time. found the root cause. not a Vault issue.

the issue was on my side. i was not clearing the wrap-ttl header (set during the wrap api call), during the unwrap call. this was causing the unwrapped response to be re-wrapped again. thanks.

**Comment by pycaster** [comment]
@jefferai @rajanadar 

I am running into a similar issue on the client side where I am trying to call unwrap API to read original response. Here is the piece of code doing that,

```
func unwrapVaultResponse(wrappedInfo *api.SecretWrapInfo) (*api.Secret, error) {
	log.Println("Unwrapping token response..")

	logWrappedInfo(wrappedInfo)

	config := api.DefaultConfig()
	client, err := api.NewClient(config)
	if err != nil {
		log.Println(err)
		return nil, err
	}

	client.SetToken(wrappedInfo.Token)
	client.SetAddress(VaultAddr)

	clientToken, err := client.Logical().Unwrap("")
	if err != nil {
		return nil, err
	}

	logSecret(clientToken)

	if err != nil {
		return nil, err
	}

	return clientToken, nil
}
```

But every time the returned clientToken is `null` and also I see no error from Unwrap() call. What am I doing incorrectly here? 

When I use the wrapped token for lookup and unwrap via other rest clients, it works as expected. I couldn't find information about the TTL header in the docs.

**Comment by jefferai** [comment]
@pycaster pass the wrapped token into the Unwrap call instead of setting it on the client.

**Comment by pycaster** [comment]
@jefferai 

I did that as well but yet the same result,

```
func unwrapVaultResponse(wrappedInfo *api.SecretWrapInfo) (*api.Secret, error) {
	log.Println("Unwrapping token response..")

	logWrappedInfo(wrappedInfo)

	config := api.DefaultConfig()
	client, err := api.NewClient(config)
	if err != nil {
		log.Println(err)
		return nil, err
	}

	client.SetAddress(VaultAddr)

	// Vault knows to unwrap the client token if the token to unwrap is empty.
	clientToken, err := client.Logical().Unwrap(wrappedInfo.Token)
	if err != nil {
		return nil, err
	}

	logSecret(clientToken)

	if err != nil {
		return nil, err
	}

	return clientToken, nil
}
```

Here is the sample `api.SecretWrapInfo` that I am trying to unwrap,

```
{
  "token": "cc950080-0d99-bd38-7b36-bc10b47323ac",
  "accessor": "36d8e900-6bdc-e661-735c-dbee78c76ff1",
  "ttl": 120,
  "creation_time": "2018-02-28T04:01:25.571689806Z",
  "creation_path": "auth/token/create",
  "wrapped_accessor": "35b0ce33-0f1e-ff82-fbba-bb449af5bd8b"
}
```

**Comment by pycaster** [comment]
Also, the errors that I am seeing while unwrapping especially with bad token to begin with, vault CLI return error saying `bad token` but unwrapping API seems to be returning null.

**Comment by jefferai** [comment]
It's not clear to me just from looking at your code what's wrong, however, if the CLI is working for you, since that's just a Go API client looking at what it's doing would probably be instructive.
